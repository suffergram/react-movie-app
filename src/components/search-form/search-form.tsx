import { FormEvent } from 'react';
import Button from '../button/button';
import useGetParams from '../../hooks/use-get-params';

export default function SearchForm() {
  const { search, setSearchParams } = useGetParams();

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const form = e.target as typeof e.target & {
      search: { value: string };
    };

    const query = form.search?.value;

    setSearchParams((params) => {
      params.set('search', query);
      return params;
    });
  };

  return (
    <form className="search-line" onSubmit={handleSubmit}>
      <input
        type="search"
        name="search"
        placeholder="What do you want to watch?"
        defaultValue={search ?? ''}
      />
      <Button type="submit" className="confirm">
        search
      </Button>
    </form>
  );
}
