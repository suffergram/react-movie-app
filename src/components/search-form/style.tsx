import styled from 'styled-components';
import { Button } from '../button/button';

export const StyledInput = styled.input`
  width: 713px;
  border: none;
  border-radius: 4px;
  background-color: rgba(35, 35, 35, 0.5);
  color: var(--color-5);
  padding-left: 15px;
  box-sizing: border-box;
  font-size: 20px;
`;

export const StyledButton = styled(Button)`
  width: 233px;
  font-size: 20px;
`;

export const StyledForm = styled.form`
  height: 57px;
  display: flex;
  gap: 14px;
`;
