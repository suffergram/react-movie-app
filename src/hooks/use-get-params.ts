import { useSearchParams } from 'react-router-dom';
import SearchParam, { OFFSET } from '../types/search-param';

export default function useGetParams() {
  const [searchParams, setSearchParams] = useSearchParams();

  // return {
  //   [SearchParam.Search]: searchParams.get(SearchParam.Search),
  //   [SearchParam.SortBy]: searchParams.get(SearchParam.SortBy),
  //   [SearchParam.Filter]: searchParams.get(SearchParam.Filter),
  //   [SearchParam.Offset]: searchParams.get(SearchParam.Offset),
  //   setSearchParams,
  // };

  const search = searchParams.get(SearchParam.Search) ?? '';

  const sortBy = searchParams.get(SearchParam.SortBy) ?? 'release_date';

  const filter = searchParams.get(SearchParam.Filter) ?? 'all';

  const offset = searchParams.get(SearchParam.Offset) ?? '0';

  return { search, sortBy, filter, offset, setSearchParams };
}
