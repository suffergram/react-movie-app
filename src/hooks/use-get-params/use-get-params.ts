import { useSearchParams, useRouter, usePathname } from 'next/navigation';
import { SearchParam } from '../../types/search-param';

export type SearchParamsType = Partial<Record<SearchParam, string>>;

export function useGetParams() {
  const searchParams = useSearchParams();
  const router = useRouter();
  const pathname = usePathname();

  const setSearchParams = (newParams: URLSearchParams) => {
    const routerString = `?${newParams.toString()}`;
    router.replace(`${pathname}${routerString}`);
  };

  const params: SearchParamsType = {};

  if (
    searchParams.has(SearchParam.Search) &&
    searchParams.get(SearchParam.Search) !== ''
  ) {
    params[SearchParam.Search] = searchParams.get(SearchParam.Search) ?? '';
  }

  if (
    searchParams.has(SearchParam.Filter) &&
    searchParams.get(SearchParam.Filter) !== 'all'
  ) {
    params[SearchParam.Filter] = searchParams.get(SearchParam.Filter) ?? '';
  }

  if (
    searchParams.has(SearchParam.SortBy) &&
    searchParams.get(SearchParam.SortBy) !== 'release_date'
  ) {
    params[SearchParam.SortBy] = searchParams.get(SearchParam.SortBy) ?? '';
  }

  if (
    searchParams.has(SearchParam.Offset) &&
    searchParams.get(SearchParam.Offset) !== '0'
  ) {
    params[SearchParam.Offset] = searchParams.get(SearchParam.Offset) ?? '';
  }

  return [params, setSearchParams] as const;
}
