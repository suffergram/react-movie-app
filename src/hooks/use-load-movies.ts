import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import RootState from '../types/root-state';
import fetchMovies from '../state/fetch-movies';

export default function useLoadMovies() {
  const dispatch: ThunkDispatch<RootState, unknown, AnyAction> = useDispatch();

  const { movies, isLoading, error, sort, filter, totalAmount, offset } =
    useSelector((state: RootState) => state.movieState);

  useEffect(() => {
    dispatch(fetchMovies(filter, sort, offset));
  }, [dispatch, sort, filter, offset]);

  return { isLoading, error, movies, totalAmount };
}
