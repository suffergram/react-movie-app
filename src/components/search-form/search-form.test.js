import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SearchForm } from './search-form';
import { useGetParams } from '../../hooks/use-get-params';

jest.mock('../../hooks/use-get-params', () => ({
  useGetParams: jest.fn(() => [{}]),
}));

describe('Search input', () => {
  it('Matches the snapshot', () => {
    render(
      <MemoryRouter>
        <SearchForm />
      </MemoryRouter>
    );

    const searchInputElement = screen.getByRole('searchbox');

    expect(searchInputElement).toMatchSnapshot();
  });
})

describe('Search button', () => {
  it('Matches the snapshot', () => {
    render(
      <MemoryRouter>
        <SearchForm />
      </MemoryRouter>
    );

    const searchButtonElement = screen.getByRole('button');

    expect(searchButtonElement).toMatchSnapshot();
  });
})

describe('Search input', () => {
  it('Can contain a value', async () => {
    render(
      <MemoryRouter>
        <SearchForm />
      </MemoryRouter>
    );

    const inputValue = 'the';
    const searchInputElement = screen.getByRole('searchbox');

    await userEvent.type(searchInputElement, inputValue);

    waitFor(() => expect(searchInputElement).toHaveValue(inputValue));
  });
})

describe('Search Form', () => {
  it('Handles input value on button click', async () => {
    const setUserSearchParamsMock = jest.fn();

    useGetParams.mockImplementation(() => [{}, setUserSearchParamsMock]);

    render(
      <MemoryRouter>
        <SearchForm />
      </MemoryRouter>
    );

    const inputValue = 'the';
    const searchInputElement = screen.getByRole('searchbox');
    const searchButtonElement = screen.getByRole('button');
    const expectedReturnValue = new URLSearchParams();

    expectedReturnValue.set('search', inputValue);

    await userEvent.type(searchInputElement, inputValue);
    await userEvent.click(searchButtonElement);

    expect(setUserSearchParamsMock).toHaveBeenCalledWith(expectedReturnValue);
  });
});

