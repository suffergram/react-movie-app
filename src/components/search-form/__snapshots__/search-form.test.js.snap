// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search button Matches the snapshot 1`] = `
<button
  class="sc-beyTiQ iwJzVu sc-dmyDmy hVEAxv"
  type="submit"
>
  search
</button>
`;

exports[`Search input Matches the snapshot 1`] = `
<input
  class="sc-guDMob hntGaG"
  name="search"
  placeholder="What do you want to watch?"
  type="search"
/>
`;
